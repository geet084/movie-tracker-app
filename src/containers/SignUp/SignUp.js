import { Link } from 'react-router-dom';
import React, { Component } from 'react';
import API from '../../utils/api';
import buildInput from '../../utils/helpers';

class SignUp extends Component {
  constructor() {
    super();
    this.state = {
      user: {
        name: '',
        email: '',
        password: '',
      },
      response: ''
    }
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    let message;
    try {
      const response = await API.postData(this.state.user, '/new');
      message = `${response.message}. Please log in.`
    } catch (error) {
      message = 'User email already exists, please try again or log in.';
    }

    this.formRef.reset()
    this.setState({
      user: {
        name: '',
        email: '',
        password: '',
      }, 
      response: message,
    })
  }

  handleChange = (e) => {
    let { name, value } = e.target;
    this.setState({
      user: {...this.state.user, [name]: value}
    });
  }

  render() {
    const { user, response } = this.state
    const inputFields = Object.keys(user).map(field => buildInput(field, this.handleChange))
    return (
      <div className='overlay-div'>
        <div className='login-div'>
          <Link to='/'>
            <button className='home-link'></button>
          </Link>
          <form autoComplete='off' onSubmit={this.handleSubmit} ref={(el) => this.formRef = el}>
            {inputFields}
            <input className='submit-button' type="submit" />
          </form>
          <Link className='sign-up-login-link' to='/login'>Login</Link>
          <p className='response'>{response}</p>
        </div>
      </div>
    )
  }
}

export default SignUp;