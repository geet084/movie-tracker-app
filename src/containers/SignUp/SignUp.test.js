import React from 'react'
import { shallow } from 'enzyme'
import SignUp from './SignUp'
import API from '../../utils/api'

describe('Signup', () => {
  let wrapper;

  describe('defaults', () => {
    beforeEach(() => {
      wrapper = shallow(
        <SignUp />
      )
    })
    
    it('should match snapshot', () => {

      expect(wrapper).toMatchSnapshot();
    })

    it('should have a default state', () => {
      const expected = {
        user: { name: '', email: '', password: '' },
        response: ''
      }

      expect(wrapper.state()).toEqual(expected)
    })
  })

  describe('handleChange', () => {
    it('should update state with the input value', () => {
      const mockEvent = {
        target: { name: 'name', value: 'new value' }
      }
      const expected = {
        name: 'new value', email: '', password: ''
      }

      wrapper.instance().handleChange(mockEvent)

      expect(wrapper.state('user')).toEqual(expected)
    })
  })

  describe('handleSubmit', () => {
    it('should sucessfully handle a new user signup', async () => {
      const mockEvent = { preventDefault: () => { } }
      const mockData = {
        id: 1,
        message: "New user created",
        status: "success",
      }
      API.postData = jest.fn().mockImplementation(() => {
        return mockData
      })
      const expected = 'New user created'

      await wrapper.instance().handleSubmit(mockEvent)

       expect(wrapper.state('response')).toEqual(expected)
    })

    it('should sucessfully handle a duplicate user signup', async () => {
      const mockEvent = { preventDefault: () => { } }
      const mockData = {
        error: "Key email already exists."
      }
      API.postData = jest.fn().mockImplementation(() => {
        return mockData
      })
      const expected = 'User already created'

      await wrapper.instance().handleSubmit(mockEvent)

      expect(wrapper.state('response')).toEqual(expected)
    })

    it.skip('should unsuccessfully handle user signup', () => {

    })
  })

})