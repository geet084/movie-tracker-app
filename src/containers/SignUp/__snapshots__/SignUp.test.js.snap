// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup defaults should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUp />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          replace={false}
          to="/"
        >
          HOME
        </Link>,
        <form
          onSubmit={[Function]}
        >
          <label>
            name
            <input
              name="name"
              onChange={[Function]}
              required={true}
              type="name"
            />
          </label>
          <label>
            email
            <input
              name="email"
              onChange={[Function]}
              required={true}
              type="email"
            />
          </label>
          <label>
            password
            <input
              name="password"
              onChange={[Function]}
              required={true}
              type="password"
            />
          </label>
          <input
            type="submit"
          />
        </form>,
        <h3>
          
        </h3>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "HOME",
          "replace": false,
          "to": "/",
        },
        "ref": null,
        "rendered": "HOME",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [
              <label>
                name
                <input
                  name="name"
                  onChange={[Function]}
                  required={true}
                  type="name"
                />
              </label>,
              <label>
                email
                <input
                  name="email"
                  onChange={[Function]}
                  required={true}
                  type="email"
                />
              </label>,
              <label>
                password
                <input
                  name="password"
                  onChange={[Function]}
                  required={true}
                  type="password"
                />
              </label>,
            ],
            <input
              type="submit"
            />,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "name",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "name",
                <input
                  name="name"
                  onChange={[Function]}
                  required={true}
                  type="name"
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              "name",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "name": "name",
                  "onChange": [Function],
                  "required": true,
                  "type": "name",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": "email",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "email",
                <input
                  name="email"
                  onChange={[Function]}
                  required={true}
                  type="email"
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              "email",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "name": "email",
                  "onChange": [Function],
                  "required": true,
                  "type": "email",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": "password",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "password",
                <input
                  name="password"
                  onChange={[Function]}
                  required={true}
                  type="password"
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              "password",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "name": "password",
                  "onChange": [Function],
                  "required": true,
                  "type": "password",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "h3",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            replace={false}
            to="/"
          >
            HOME
          </Link>,
          <form
            onSubmit={[Function]}
          >
            <label>
              name
              <input
                name="name"
                onChange={[Function]}
                required={true}
                type="name"
              />
            </label>
            <label>
              email
              <input
                name="email"
                onChange={[Function]}
                required={true}
                type="email"
              />
            </label>
            <label>
              password
              <input
                name="password"
                onChange={[Function]}
                required={true}
                type="password"
              />
            </label>
            <input
              type="submit"
            />
          </form>,
          <h3>
            
          </h3>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "HOME",
            "replace": false,
            "to": "/",
          },
          "ref": null,
          "rendered": "HOME",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [
                <label>
                  name
                  <input
                    name="name"
                    onChange={[Function]}
                    required={true}
                    type="name"
                  />
                </label>,
                <label>
                  email
                  <input
                    name="email"
                    onChange={[Function]}
                    required={true}
                    type="email"
                  />
                </label>,
                <label>
                  password
                  <input
                    name="password"
                    onChange={[Function]}
                    required={true}
                    type="password"
                  />
                </label>,
              ],
              <input
                type="submit"
              />,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "name",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "name",
                  <input
                    name="name"
                    onChange={[Function]}
                    required={true}
                    type="name"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                "name",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "name": "name",
                    "onChange": [Function],
                    "required": true,
                    "type": "name",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": "email",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "email",
                  <input
                    name="email"
                    onChange={[Function]}
                    required={true}
                    type="email"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                "email",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "name": "email",
                    "onChange": [Function],
                    "required": true,
                    "type": "email",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": "password",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "password",
                  <input
                    name="password"
                    onChange={[Function]}
                    required={true}
                    type="password"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                "password",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "name": "password",
                    "onChange": [Function],
                    "required": true,
                    "type": "password",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "type": "submit",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "h3",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
